src/lib/operators/aggregate.cpp

```diff
@@ -341,18 +343,40 @@ std::shared_ptr<const Table> Aggregate::_on_execute() {
   This is done by creating a vector that contains the AggregateKey for each row.
   It is gradually built by visitors, one for each group column.
   */
-  _keys_per_chunk = std::vector<std::shared_ptr<std::vector<AggregateKey>>>(input_table->chunk_count());
+
+  // Allocate a temporary memory buffer, for more details see aggregate.hpp
+  size_t needed_size_per_aggregate_key =
+      aligned_size<AggregateKey>() + _groupby_column_ids.size() * aligned_size<AggregateKeyEntry>();
+  size_t needed_size = aligned_size<KeysPerChunk>() + input_table->chunk_count() * aligned_size<AggregateKeys>() +
+                       input_table->row_count() * needed_size_per_aggregate_key;
+
+  auto temp_buffer = boost::container::pmr::monotonic_buffer_resource(needed_size);
+  auto allocator = AggregateKeysAllocator{PolymorphicAllocator<AggregateKeys>{&temp_buffer}};
+  auto start_next_buffer_size = temp_buffer.next_buffer_size();
+
+  // Create the actual data structure
+  auto keys_per_chunk = KeysPerChunk{allocator};
+  keys_per_chunk.reserve(input_table->chunk_count());
 
   for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
-    _keys_per_chunk[chunk_id] = std::make_shared<std::vector<AggregateKey>>(input_table->get_chunk(chunk_id)->size(),
-                                                                            AggregateKey(_groupby_column_ids.size()));
+    keys_per_chunk.emplace_back(input_table->get_chunk(chunk_id)->size(), AggregateKey(_groupby_column_ids.size()));
+  }
+
+  // Make sure that we did not have to allocate more memory than originally computed
+  if (temp_buffer.next_buffer_size() != start_next_buffer_size) {
+    // The buffer sizes are increasing when the current buffer is full. We can use this to make sure that we allocated
+    // enough space from the beginning on. It would be more intuitive to compare current_buffer(), but this seems to
+    // be broken in boost: https://svn.boost.org/trac10/ticket/13639#comment:1
+    PerformanceWarning(std::string("needed_size ") + std::to_string(needed_size) +
```

Comment:
This is basically a warning for us developers, but the anwender cannot do anything about it? In case we see that warning often, we are going to change something about the size calculation above?

Reply:
We could run some tests on it. I recall seeing this warning popping up from time to time (not super often) when doing the benchmarks above.

Reply:
This is surprising. I haven't seen it. Can you tell me in which configuration you are seeing the message?

Reply:
Also, which OS and which compiler?

Reply:
Ubuntu (nemea), default compiler (gcc?). It was when I did the benchmarks for this. The output is gone unfortunately, so I can't verify that now. I can run the benchmarks again and see if it pops up again?

Oh and configuration, I think it was the last one with TPCH scale 1, and the number of runs equals two times the number of used cores.

Reply:
Maybe you could keep an eye on this? Thanks to the warning, this won't get lost.

Reply:
I'll keep it in mind and see if it pops up again in future benchmarks.